src/protocol/attribution/apply_attribution_window.rs

```diff
@@ -0,0 +1,351 @@
+                    .zip(stop_bits.iter().skip(1))
+                    .enumerate()
+                    .map(|(i, ((prev, curr), b))| {
+                        let c = t_delta_context.clone();
+                        let record_id = RecordId::from(i);
+                        let delta = curr.timestamp.clone() - &prev.timestamp;
+                        async move { c.multiply(record_id, &delta, b).await }
+                    }),
+            )
+            .await?,
+        )
+        .collect::<Vec<_>>();
+
+    // Do the prefix-sum accumulation of timedeltas. This is similar to
+    // `do_the_binary_tree_thing()`, but it sums up values in top-down order.
+    for (depth, step_size) in std::iter::successors(Some(1_usize), |prev| prev.checked_mul(2))
+        .take_while(|&v| v < num_rows)
+        .enumerate()
+    {
+        let record_count = num_rows - step_size;
+
+        let depth_i_ctx = ctx
+            .narrow(&InteractionPatternStep::from(depth))
+            .set_total_records(record_count);
+        let mt_ctx = depth_i_ctx.narrow(&Step::PreviousStopBitTimesCurrent);
+
+        let mut futures = Vec::with_capacity(record_count);
+
+        // `i` starts from 0 for RecordId. The actual logic skips the first
+        // `step_size` rows.
+        for i in 0..record_count {
+            let index = i + step_size;
+
+            let c1 = depth_i_ctx.clone();
+            let c2 = mt_ctx.clone();
+            let record_id = RecordId::from(i);
+
+            let previous_stop_bit = &stop_bits[index - step_size];
+            let current_stop_bit = &stop_bits[index];
+            let previous_t_delta = &t_delta[index - step_size];
+
+            futures.push(async move {
+                try_join(
+                    // if the current stop_bit is 0, don't accumulate
+                    c1.multiply(record_id, previous_t_delta, current_stop_bit),
+                    // next `stop_bit`
+                    c2.multiply(record_id, previous_stop_bit, current_stop_bit),
+                )
+                .await
+            });
+        }
+
+        let results = try_join_all(futures).await?;
+
+        results
+            .into_iter()
+            .enumerate()
+            .for_each(|(i, (delta, stop_bit))| {
+                t_delta[i + step_size] += &delta;
+                stop_bits[i + step_size] = stop_bit;
+            });
+    }
+
+    Ok(t_delta)
+}
+
+/// Creates a vector of trigger values where values are set to `0` if the time delta
+/// from their nearest source event exceed the specified attribution window cap.
+///
+/// This protocol executes the bit-decomposition protocol in order to compare shares
+/// of time deltas in `F`.
+///
+/// # Errors
+/// Fails if the bit-decomposition, bitwise comparison, or multiplication fails.
+async fn zero_out_expired_trigger_values<F, C, T>(
+    ctx: &C,
+    input: &[MCApplyAttributionWindowInputRow<F, T>],
+    time_delta: &[T],
+    cap: u32,
+) -> Result<Vec<T>, Error>
```

Comment:
you may want to talk to @richajaindce as she recently put a lot of effort refactoring these allocations for protocols that process input rows one by one. This seems like another good use case for iterators that could simplify our life in the future. As a bonus point it should allow zipping `time_delta` and `input` together with no runtime overhead

Reply:
Ah, right. I noticed the change after I started to work on this. Will change them to iterators where possible.

Reply:
We can do this in a follow-up diff.